### DISCUSSION TITLE
What do you all think about more charts?

### DISCUSSION THREAD

**Discussion by mwufi** [excalidraw/excalidraw#2533] [Q&A] [answered]
Is it worth the complexity to add more kinds of charts?

- Line charts
![image](https://user-images.githubusercontent.com/30219253/102025346-dbde5a00-3d5c-11eb-8ae8-c5050d4a9910.png)

- Pie charts

- Bar charts (default)

And then we might add a little menu sidebar to switch between the types.

Implementation-wise, there's a few options
1) store the raw data (and allow switching of chart types)
2) ask people to pick which type of chart they'd like (as a modal?) **before** inserting the elements

**Comment by lipis** [comment]
While the line chart is indeed interesting, the pie chart is another beast, as we need to have labels with all the different colors and we can easily run out if the chart has more than 14 rows (our default number of bg colors).

Having said that, I think we don't want to overcomplicate things and asking users what to paste or storing metadata with the graph to change it later, as we don't have anything in place to support it.

**Reply by dwelle** [reply]
One way would be to simply have a contextmenu item "paste special..." which would open a submenu where we could list "paste as bar chart", "paste as line chart"...:

![image](https://user-images.githubusercontent.com/5153846/102088936-deec4f80-3e1b-11eb-96f9-e444b5a8e613.png)

We don't need to support converting an existing chart type into a different chart type, so no need for storing metadata.

Of course, the "complexity" argument stands.

**Reply by lipis** [reply]
I also thought about that.. but we don't know if the contents of the clipboard is data or not to show it only there. We get access to clipboard when the user presses paste.

**Reply by dwelle** [reply]
With the new Clipboard API we can read from it e.g. when user righ-clicks to invoke the menu. Though for new users who didn't yet gave the permission it would prompt them on contextmenu invocation which isn't ideal :(.

**Reply by dwelle** [reply]
We can also check if user already granted permissions and only show that "paste special" if granted:

```
await navigator.permissions.query({ name: 'clipboard-read', allowWithoutGesture: false });
```

Though that has its own UX problems :), not to mention that the clipboard API only works on Chrome (don't recall if FF allows reading from it).

**Comment by andyczerwonka** [comment]
The existing function is **really** cool, but unless it's extended to handle some basic charts with at least one level grouping, it's really just a novelty that I don't think will be used much. I would definitely go with option 1 above.

**Reply by dwelle** [reply]
I don't think storing metadata is the way to go. It would become stale after you made manual changes.

If people want a different chart, they'll just have to paste the data again.

**Comment by petehunt** [comment]
yes i would definitely like to see more charts. i think option 2 above is easiest to implement.

when i originally built the mvp of this i had a version that did pie charts too, but we didn't have sharp corners back then so they looked bad.

when building dataviz at work we really only use two types of charts often:
* line charts
* donut charts (these are supposed to be less misleading than pie charts)

i think a good next addition to this feature would be the ability to pick between a bar chart, a single-line chart and a donut chart while pasting. in the future we could improve on this feature to support multiple series (and enable things like stacked bars, mountain charts and multi-line charts) but i think it's critical that implementation complexity and UI complexity of this thing remains low.

i'm +1 on @dwelle's suggestion that we shouldn't store metadata. one of the things that i really like about this feature is you can modify the individual elements of the chart with no restrictions.

**Comment by lipis** [comment]
**TLDR;** After importing a chart, provide a way to change it to other types with a timeout.

-----

So here is an idea: Right after the successful import of the chart we display some sort of a _notification_ somewhere on the chart so it will be easily changed to a line chart (and/or possibly more kind of charts).

<img width="459" alt="Screenshot 2020-12-22 at 1 01 17 AM" src="https://user-images.githubusercontent.com/125676/102830076-3462c680-43f1-11eb-8f4a-9863f2d3fd54.png">

This _notification_ will appear only after a paste (like a [En_passant](https://en.wikipedia.org/wiki/En_passant) in chess) and will start fading out after 4 seconds until it will completely disappears. If they missed it it's fine. They can just paste again and get it back. It's a hidden feature anyway. We can tweak the style and the timeout, so it will it will as pleasant as possible :)

After the user pressed the `Line chart`, we delete the inserted chart and insert a `Line chart` and now showing this _notification_ again to switch it to a `Bar chart`. If nothing happens it will fade away and so on...

People are not inserting too many charts that often and we'll give them a quick chance to switch it to something else. Maybe storing it in app state just the last pasted chart so next time it will be the same as the last one.

### Implementation wise

- Add `chartType`  to `AppState`, with default set to `bar` (possible types: `bar`, `line`, `pie`, etc)
- When pasting charts always paste the `chartType` that is set in `appState`
- Give user a choice of all other types other than the current `chartType` and update it in `appState`
- Set a timeout to slowly fade out the options
- This option should be also saved in a file so if you paste again a chart in that document it will match the last choice
- It's a hidden feature so we don't have to have anything changed in the UI

**Reply by lipis** [reply]
I opened a PR for ideas: https://github.com/excalidraw/excalidraw/pull/2658

**Comment by lipis** [comment, marked as answer]
We just released the line charts! https://twitter.com/excalidraw/status/1343232547543797760

and we have a foundation for more..

**Reply by OrenMe** [reply]
Beautiful, I like the preview selection

**Comment by kinthaiofficial** [comment]
More chart types in Excalidraw is an interesting direction — it sits at an interesting intersection between "whiteboard" and "diagramming tool" that most competing tools don't occupy well.

**The use case gap Excalidraw charts fill:**
Meeting sketches where you want to quickly show "here's roughly how the data looks" without the overhead of exporting from a real charting library. The hand-drawn aesthetic actually helps signal "this is exploratory" rather than "this is authoritative data."

**What would make chart support valuable:**

**Data-bound charts (reactive)**: Link a chart to a data source (inline JSON, CSV, or a formula). When the data changes, the chart updates. Without this, charts are just decorative shapes.

**Diagram-chart hybrids**: Excalidraw's strength is annotations + drawings alongside charts. A bar chart with hand-drawn arrows pointing to interesting data points and text labels is more expressive than a pure chart tool.

**Export-compatible with chart tools**: If I sketch a chart in Excalidraw, I want to be able to export the underlying data in a format that Mermaid, Vega, or d3 can consume for the final version.

**LLM-generated chart data**: "Draw me a rough bar chart of Q1 revenue by region" as an Excalidraw prompt → AI generates the shape layout → you edit manually. This would be a compelling use case for AI integration.

**The concern:**
Adding complex chart editing would compete with Excalidraw's core simplicity. The sketch-first philosophy is a feature, not a limitation. Charts that require precise data entry lose the sketch feel. Bar for adding chart support: must feel as fast and approximate as drawing a box.
